pkg/restore/restore.go

```diff
@@ -373,103 +381,89 @@ func (ctx *restoreContext) execute() (Result, Result) {
+						ctx.log.WithError(errors.WithStack((err))).Warn("Got error trying to update restore's status.progress")
+					}
+					lastUpdate = nil
+				}
 			}
+		}
+	}()
 
-			// get target namespace to restore into, if different
-			// from source namespace
-			targetNamespace := namespace
-			if target, ok := ctx.restore.Spec.NamespaceMapping[namespace]; ok {
-				targetNamespace = target
-			}
+	totalItems, i, existingNamespaces := 0, 0, sets.NewString()
+
+	for _, selectedResource := range selectedResourceCollection {
+		totalItems += selectedResource.totalItems
+	}
+	for _, selectedResource := range selectedResourceCollection {
+		groupResource := schema.ParseGroupResource(selectedResource.resource)
+
+		for namespace, selectedItems := range selectedResource.selectedItemsByNamespace {
+			for _, selectedItem := range selectedItems {
+				// if we don't know whether this namespace exists yet, attempt to create
+				// it in order to ensure it exists. Try to get it from the backup tarball
+				// (in order to get any backed-up metadata), but if we don't find it there,
+				// create a blank one.
+				if namespace != "" && !existingNamespaces.Has(selectedItem.targetNamespace) {
+					logger := ctx.log.WithField("namespace", namespace)
+					ns := getNamespace(logger, archive.GetItemFilePath(ctx.restoreDir, "namespaces", "", namespace), selectedItem.targetNamespace)
+					if _, err := kube.EnsureNamespaceExistsAndIsReady(ns, ctx.namespaceClient, ctx.resourceTerminatingTimeout); err != nil {
+						errs.AddVeleroError(err)
+						continue
+					}
 
-			// if we don't know whether this namespace exists yet, attempt to create
-			// it in order to ensure it exists. Try to get it from the backup tarball
-			// (in order to get any backed-up metadata), but if we don't find it there,
-			// create a blank one.
-			if namespace != "" && !existingNamespaces.Has(targetNamespace) {
-				logger := ctx.log.WithField("namespace", namespace)
-				ns := getNamespace(logger, archive.GetItemFilePath(ctx.restoreDir, "namespaces", "", namespace), targetNamespace)
-				if _, err := kube.EnsureNamespaceExistsAndIsReady(ns, ctx.namespaceClient, ctx.resourceTerminatingTimeout); err != nil {
-					errs.AddVeleroError(err)
+					// keep track of namespaces that we know exist so we don't
+					// have to try to create them multiple times
+					existingNamespaces.Insert(selectedItem.targetNamespace)
+				}
+				obj, err := archive.Unmarshal(ctx.fileSystem, selectedItem.path)
+				if err != nil {
+					errs.Add(selectedItem.targetNamespace, fmt.Errorf("error decoding %q: %v", strings.Replace(selectedItem.path, ctx.restoreDir+"/", "", -1), err))
 					continue
 				}
-
-				// keep track of namespaces that we know exist so we don't
-				// have to try to create them multiple times
-				existingNamespaces.Insert(targetNamespace)
+				w, e := ctx.restoreItem(obj, groupResource, selectedItem.targetNamespace)
+				actualTotalItems := len(ctx.restoredItems) + (totalItems - (i + 1))
```

Comment:
I'm unclear as to why we're doing `(totalItems - (i+1)`) here. Can you clarify with a comment?

Reply:
@nrb since there may be additional items restored by the plugins, the initial count `totalItems` needs to be adjusted to include those additional items. `ctx.restoredItems` has all the items actually restored. But we don't want to count items twice. Therefore, this adjustment is required.

Reply:
Ahh, ok. Can you add that as a code comment, please?